### DISCUSSION TITLE
How to apply effetcs to a mp3 stream?

### DISCUSSION THREAD

**Discussion by KatePimbleton** [pschatzmann/arduino-audio-tools#662] [Q&A] [answered]
Hi,

I followed the exemple on how to apply effects to any stream, though
I can't find a way to apply it on a mp3 read from a sdCard. Is it implemented
yet?

Thank you

**Comment by pschatzmann** [comment, marked as answer]
Sure, you would just decode the mp3 first, how to do this is described in the documentation

**Reply by KatePimbleton** [reply]
I'm able to decode the mp3, following the documentation, then I tried to apply an effect to the AudioPlayer:

MP3DecoderHelix decoder; 
AudioPlayer player(source, kit, decoder);
AudioEffectStream effects(player);

And it seems I'm totally wrong...

**Comment by KatePimbleton** [comment]
following the example: " streams-audiokit-effects-audiokit" I tried to add effects to the "player-sd-audiokit" exemple.
With no luck, any help welcome!

this is what I came up with so far:

#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioSourceSD.h" // or AudioSourceIdxSD.h
#include "AudioCodecs/CodecMP3Helix.h"

const int sample_rate = 44100;
const char *startFilePath = "/";
const char* ext = "mp3";

Delay del(0, sample_rate);

AudioSourceSD source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
AudioKitStream kit;
MP3DecoderHelix decoder;  // or change to MP3DecoderMAD
AudioPlayer player(source, kit, decoder);

AudioEffectStream effects(kit);
StreamCopy copier(kit, effects);

void updateValues() {
  // update values in controls
  del.setActive(1);
  del.setDepth(1);
  del.setFeedback(0.9);
  del.setDuration(1000);
}
void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sd_active = true;
  kit.begin(cfg);

  effects.addEffect(del);
  effects.begin(cfg);
  updateValues();

  // setup player
  player.setVolume(0.7);
  player.begin();

}
void loop() {
  copier.copy();
  player.copy();
  kit.processActions();
}

Cheers

**Comment by pschatzmann** [comment]
You got the processing chain completly wrong: if you want to use the AudioPlayer, you just send the output the the effects instead of the kit. No further data copy is necessary! : AudioPlayer player(source, effects, decoder);

**Reply by KatePimbleton** [reply]
Thank you! It worked, the audio is stuttering though. 
Is there something else I'm missing ?

I'm playing mp3 files @ 224kb/s 
I use the ESP32-audiokit "type 5" ,

here is the code used:



#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioSourceSD.h" // or AudioSourceIdxSD.h
#include "AudioCodecs/CodecMP3Helix.h"

const int sample_rate = 44100;
const char *startFilePath = "/";
const char* ext = "mp3";

Delay del(0, sample_rate);

AudioSourceSD source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
AudioKitStream kit;
MP3DecoderHelix decoder;  // or change to MP3DecoderMAD
AudioEffectStream effects(kit);
AudioPlayer player(source, effects, decoder);

void updateValues() {
  // update values in controls
  del.setActive(1);
  del.setDepth(0.5);
  del.setFeedback(0.5);
  del.setDuration(500);
}
void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sd_active = true;
  kit.begin(cfg);
  effects.addEffect(del);
  effects.begin(cfg);
  updateValues();
  // setup player
  player.setVolume(1);
  player.begin();
}
void loop() {
  player.copy();
  kit.processActions();
}

**Comment by pschatzmann** [comment]
Did you try to set the log level to warning ?

**Reply by KatePimbleton** [reply]
I set it to warning, once in a while this message appears:

"[W] AudioPlayer.h : 445 - -> timeout - moving by 1"

**Comment by pschatzmann** [comment]
Updated the code to change this to a Info message instead of a warning. It just indicated that the file has finished and it's moving to the next file now...

**Comment by KatePimbleton** [comment]
Hi,

I tried applying effects on a short "wav" file ( 16BIT 44100hz )
instead of mp3, the result is the same, very loud crakling.
I also tested on a "type 7" board, with no luck.

Any clue about what parameter I could try to tweak ?


code used:


#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioSourceSD.h" // or AudioSourceIdxSD.h
#include "AudioCodecs/CodecWAV.h"

const char *startFilePath = "/";
const char* ext = "wav";
const int sample_rate = 44100;

Delay del(0, sample_rate);

AudioSourceSD source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
AudioKitStream kit;
WAVDecoder decoder; 
AudioEffectStream effects(kit);
AudioPlayer player(source, effects, decoder);

void updateValues() {
  del.setActive(1);
  del.setDepth(0.5);
  del.setFeedback(0.5);
  del.setDuration(500);
}
void setup() {
  // Serial.begin(115200);
   AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sd_active = true;
  kit.begin(cfg);
  kit.setVolume(100);

  effects.addEffect(del);
  effects.begin(cfg);
  updateValues();
  // setup player
  player.setVolume(0.7);
  player.begin();
}
void loop() {
  player.copy();
  kit.processActions();
}

**Comment by pschatzmann** [comment]
Not sure, where this could come from. 
- Did you try some other effects ?
- Did you try to Comment out the processActions()

**Reply by KatePimbleton** [reply]
I Tried with "distortion" with  and commented out "processActions()",
I still hear loud artifacts.

When I apply effects on a microphone stream it works well.

**Comment by KatePimbleton** [comment]
I configure channel/sample.rate in the setup :

 cfg.channels = 1;
 cfg.sample_rate = 44100;

and the artifacts disappeared.
